### PR TITLE
Allow `reject_transition` to use current state

### PR DESCRIPTION
**Pull request by madkinsz**
Co-authored with @anticorrelator 

Previously a REJECT response required a new state which was then always written to the database. Here, the `OrchestrationRule.reject_transition` utility is updated to accept `state=None`. When this is used, the REJECT orchestration response will contain the current state of the run and a new state will not be appended to the run's states. This allows REJECT semantics to be used to prevent transition to a proposed state without providing an explicit alternative. This is useful when a client is proposing a state with the assumption that they are in a different starting state than the orchestrator has.

Required for #7738 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/orion/orchestration/rules.py [hunk 1]
```diff
@@ -263,7 +265,12 @@ async def _validate_proposed_state(
         self,
         db: OrionDBInterface,
     ):
-        if self.proposed_state is not None:
+        if (
+            self.response_status == SetStateStatus.REJECT
+            and self.proposed_state is None
+        ):
+            validated_orm_state = self.run.state
```
Comment (madkinsz) [inline src/prefect/orion/orchestration/rules.py:272]: Could we always do this if the proposed state is null? It seems nice to have it not just apply to REJECT semantics. Or do we perform logic based on `self.validated_state is None` later?

**Comment by anticorrelator** [comment]
this idea is kickass, it's a small change that really allows us to be more expressive in our communication with clients in a backwards compatible way

File: src/prefect/orion/orchestration/rules.py (hunk 1 above)
Comment (anticorrelator) [reply src/prefect/orion/orchestration/rules.py:273]: I think we can! If our state machine acts like a markov chain, if orchestration doesn't specify a state only the current state makes sense as information

**Review by anticorrelator — APPROVED** [review]

**Review by chrisguidry — APPROVED** [review]
This all makes sense to me!

**Pull request merged**